ocaml/utils/compilation_unit.ml

```diff
@@ -123,146 +139,160 @@ end = struct
+   numbers. *)
+type with_prefix =
+  { name : Name.t;
+    for_pack_prefix : Prefix.t
+  }
+
+(* type t = Without_prefix of Name.t [@@unboxed] | With_prefix of with_prefix *)
+type t = Obj.t
+
+(* Some manual inlining is done here to ensure good performance under
+   Closure. *)
+
+let for_pack_prefix_and_name t =
+  let tag = Obj.tag t in
+  assert (tag = 0 || tag = Obj.string_tag);
+  if tag <> 0
+  then Prefix.empty, Sys.opaque_identity (Obj.obj t : Name.t)
+  else
+    let with_prefix = Sys.opaque_identity (Obj.obj t : with_prefix) in
+    with_prefix.for_pack_prefix, with_prefix.name
+
+let name t =
+  let tag = Obj.tag t in
+  assert (tag = 0 || tag = Obj.string_tag);
+  if tag <> 0
+  then Sys.opaque_identity (Obj.obj t : Name.t)
+  else
+    let with_prefix = Sys.opaque_identity (Obj.obj t : with_prefix) in
+    with_prefix.name
+
+let for_pack_prefix t =
+  let tag = Obj.tag t in
+  assert (tag = 0 || tag = Obj.string_tag);
+  if tag <> 0
+  then Prefix.empty
+  else
+    let with_prefix = Sys.opaque_identity (Obj.obj t : with_prefix) in
+    with_prefix.for_pack_prefix
 
 let create for_pack_prefix name =
-  if not (Prefix.is_empty for_pack_prefix) then begin
+  if not (Prefix.is_empty for_pack_prefix)
+  then (
     Name.check_as_path_component name;
     ListLabels.iter ~f:Name.check_as_path_component
-      (for_pack_prefix |> Prefix.to_list)
-  end;
-  { name;
-    for_pack_prefix;
-    hash = Hashtbl.hash (name, for_pack_prefix)
-  }
+      (for_pack_prefix |> Prefix.to_list));
+  if Prefix.is_empty for_pack_prefix
+  then Sys.opaque_identity (Obj.repr name)
+  else Sys.opaque_identity (Obj.repr { for_pack_prefix; name })
 
-let create_child parent name =
+let create_child parent name_ =
   let prefix =
-    (parent.for_pack_prefix |> Prefix.to_list) @ [ parent.name ]
-    |> Prefix.of_list
+    (for_pack_prefix parent |> Prefix.to_list) @ [name parent] |> Prefix.of_list
   in
-  create prefix name
+  create prefix name_
 
 let of_string str =
   let for_pack_prefix, name =
     match String.rindex_opt str '.' with
     | None -> Prefix.empty, Name.of_string str
     | Some 0 ->
-        (* See [Name.check_as_path_component]; this allows ".cinaps" as a
-           compilation unit *)
-        Prefix.empty, Name.of_string str
-    | Some _ ->
-        Misc.fatal_errorf "[of_string] does not parse qualified names"
+      (* See [Name.check_as_path_component]; this allows ".cinaps" as a
+         compilation unit *)
+      Prefix.empty, Name.of_string str
+    | Some _ -> Misc.fatal_errorf "[of_string] does not parse qualified names"
   in
   create for_pack_prefix name
 
 let dummy = create Prefix.empty (Name.of_string "*none*")
 
 let predef_exn = create Prefix.empty Name.predef_exn
 
-let name t = t.name
-
 let name_as_string t = name t |> Name.to_string
 
-let for_pack_prefix t = t.for_pack_prefix
+let with_for_pack_prefix t for_pack_prefix = create for_pack_prefix (name t)
```

Comment:
(Now checks whether the passed prefix
is valid, probably a good thing. Nothing
in the code seems to try and build an
invalid prefix - I am thinking of something
like `*predef*`.)

Reply:
This indeed should be ok.